### PR TITLE
Add Encoding Span-based APIs

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @KrzysztofCwalina, @tarekgh, @jkotas 
Contributes to https://github.com/dotnet/corefx/issues/22431

### CODE REVIEW

File: src/mscorlib/shared/System/Text/UTF8Encoding.cs [hunk 1]
```diff
@@ -2496,7 +2498,8 @@ public override byte[] GetPreamble()
             else
                 return Array.Empty<byte>();
         }
-
+        
+        public override ReadOnlySpan<byte> Preamble => _emitUTF8Identifier ? s_preamble : Array.Empty<byte>();
```
Comment (jkotas) [inline src/mscorlib/shared/System/Text/UTF8Encoding.cs]: If somebody inherited from UTF8Encoding and overridden `GetPreamble()` method, the override is not going to be respected for this property. Maybe worth a comment at least - if we consider it acceptable.

File: src/mscorlib/shared/System/Text/UTF8Encoding.cs [hunk 2]
```diff
@@ -2496,7 +2498,8 @@ public override byte[] GetPreamble()
...
 
         public override bool Equals(Object value)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Text/UTF8Encoding.cs]: We may want to get a tracking issue to implement the Span-based overloads on UTF8EncodingSealed that avoids layers of the virtual calls and argument checks.

File: src/mscorlib/shared/System/Text/UTF8Encoding.cs (hunk 1 above)
Comment (stephentoub) [inline src/mscorlib/shared/System/Text/UTF8Encoding.cs]: That's a good point.  I'll fix it.

File: src/mscorlib/shared/System/Text/UTF8Encoding.cs (hunk 2 above)
Comment (stephentoub) [inline src/mscorlib/shared/System/Text/UTF8Encoding.cs]: I'll just do it as part of this PR. Thanks.

**Comment by jkotas** [comment]
BTW: Do we plan to add the Span-based overloads for `System.Text.Encoder/Decoder` as well?

**Comment by stephentoub** [comment]
> BTW: Do we plan to add the Span-based overloads for System.Text.Encoder/Decoder as well?

I didn't look at doing so, but I can and will submit a separate API review issue.

File: src/mscorlib/shared/System/Text/UTF8Encoding.cs (hunk 2 above)
Comment (stephentoub) [inline src/mscorlib/shared/System/Text/UTF8Encoding.cs]: Actually, I'll take your suggesiton and just open a tracking issue for now.

**Pull request merged**